### ISSUE TITLE
Enable ingress conformance tests

### ISSUE THREAD

**Issue by michmike** [projectcontour/contour#2051]
Once Ingress v1 support is complete, we should enable the ingress conformance tests in CI

**Comment by jpeach** [comment]
Prerequisite for #1944

**Comment by jpeach** [comment]
Blocked by #403

**Comment by jpeach** [comment]
See: https://github.com/kubernetes-sigs/ingress-controller-conformance

And a separate effort: https://github.com/aledbf/ingress-conformance

**Comment by jpeach** [comment]
Ingress conformance is basically blocked. There's no agreement upstream on what the harness should look like, so verification checks can't be committed.

**Comment by jpeach** [comment]
Removing this from the prioritize backlog because it is blocked upstream.

**Comment by michmike** [comment]
@jpeach i would prefer we add the blocked tag and leave it in the backlog. we still want to do it at the current priority or as soon as possible given the upstream block. can you please make the changes?

**Comment by jpeach** [comment]
> @jpeach i would prefer we add the blocked tag and leave it in the backlog. we still want to do it at the current priority or as soon as possible given the upstream block. can you please make the changes?

I can move out back, but I don't see the benefit of having issues in the prioritized queue that can't be worked on.

**Comment by jpeach** [comment]
> > @jpeach i would prefer we add the blocked tag and leave it in the backlog. we still want to do it at the current priority or as soon as possible given the upstream block. can you please make the changes?
> 
> I can move it back, but I don't see the benefit of having issues in the prioritized queue that can't be worked on.

**Comment by sunjayBhatia** [comment]
Does this still look blocked given the state of https://github.com/kubernetes-sigs/ingress-controller-conformance? (theres some example of running conformance with sonobuoy and kind etc. https://github.com/kubernetes-sigs/ingress-controller-conformance/blob/master/examples/kind/.github/workflows/main.yaml)

**Comment by jpeach** [comment]
> Does this still look blocked given the state of https://github.com/kubernetes-sigs/ingress-controller-conformance? (theres some example of running conformance with sonobuoy and kind etc. https://github.com/kubernetes-sigs/ingress-controller-conformance/blob/master/examples/kind/.github/workflows/main.yaml)

See https://github.com/projectcontour/contour/issues/1944

**Comment by sunjayBhatia** [comment]
Test results from a couple runs:
[ingress-conformance-run-1.tar.gz](https://github.com/projectcontour/contour/files/5674899/ingress-conformance-run-1.tar.gz)
[ingress-conformance-run-1.tar.gz](https://github.com/projectcontour/contour/files/5674905/ingress-conformance-run-1.tar.gz)

**Comment by jpeach** [comment]
@sunjayBhatia IIRC at least some of these tests are for ingress V1, which isn't implemented in Contour.

**Comment by sunjayBhatia** [comment]
I see mention of a design doc here: https://github.com/projectcontour/contour/issues/2139, is this in flight/somewhere I can see/help with? cc @youngnick

**Comment by youngnick** [comment]
I updated #2139 with some further info, basically, the design for IngressClass is very similar to GatewayClass (the object), which is why I was initially suggesting I do them both. Happy to collaborate on something if you are interested @sunjayBhatia. Ping me and we can discuss in a high-bandwidth way somewhere.

**Comment by sunjayBhatia** [comment]
Relevant issue https://github.com/kubernetes-sigs/ingress-controller-conformance/issues/87 along with https://github.com/kubernetes-sigs/service-apis/pull/516

**Comment by xaleeks** [comment]
I guess you’re working on this already @sunjayBhatia ?  tagging it to track it

**Comment by sunjayBhatia** [comment]
This seems like the long pole here since no-one is responding: https://github.com/kubernetes-sigs/ingress-controller-conformance/pull/88

**Comment by youngnick** [comment]
@sunjayBhatia, I'd recommend just pinging Bowei on Kubernetes Slack about this one, since it's a one-liner.

**Comment by youngnick** [comment]
Seems like this is another victim of the underspecced nature of the Ingress API. If only there was some replacement that didn't have the same problems! 😝 

More seriously, I until we have ACK support, we don't have any way to check if the programming is complete, and even if we did, there's no way to tell from the object (since there's no Conditions). So I think we'll have to just change the tests to retry some number of times or wait. Can you see what other implementations do for their tests here?

**Comment by sunjayBhatia** [comment]
A quick run through and I don't see anything in CI for other Ingress controllers, but I didn't do the most digging

**Comment by youngnick** [comment]
I think the most important thing is that Contour can pass the conformance tests eventually, once things are stable. So retries or longer timeouts are okay in my book.

**Comment by sunjayBhatia** [comment]
https://github.com/kubernetes-sigs/ingress-controller-conformance/issues/92

**Comment by sunjayBhatia** [comment]
waiting on upstream ^ will make a PR but the repo has not been very active so even if I do it probably won't make it for a while

 should probably pull this out of 1.16
